File: 20. Ch-5-3.java
import java.util.*;
class Node 
{
    int key;
    int value;
    Node next;
    Node(int key, int value) 
    {
        this.key = key;
        this.value = value;
        this.next = null;
    }
}

class HashTable 
{
    int SIZE = 10;
    Node[] table;
    HashTable() 
    {
        table = new Node[SIZE];
    }

    int hashFunc(int key) 
    {
        return key % SIZE;
    }

    public void insert(int key, int value) 
    {
        int index = hashFunc(key);
        Node newNode = new Node(key, value);
        newNode.next = table[index];
        table[index] = newNode;

        System.out.println("Inserted: (" + key + "," + value + ")");
    }

    public void delete(int key) 
    {
        int index = hashFunc(key);
        Node curr = table[index];
        Node prev = null;
        while (curr != null) 
        {
            if (curr.key == key) 
            {
                if (prev == null) 
                {
                    table[index] = curr.next;
                } 
                else 
                {
                    prev.next = curr.next;
                }
                curr.next = null;
                System.out.println("Key " + key + " deleted successfully.");
                return;
            }
            prev = curr;
            curr = curr.next;
        }
        System.out.println("Key " + key + " not found.");
    }

    public void retrieve(int key) 
    {
        int index = hashFunc(key);
        Node curr = table[index];
        while (curr != null) 
        {
            if (curr.key == key) 
            {
                System.out.println(key + " -> " + curr.value);
                return;
            }
            curr = curr.next;
        }
        System.out.println("Key " + key + " not found.");
    }

    public void display() 
    {
        for (int i = 0; i < SIZE; i++) 
        {
            if (table[i] != null) 
            {
                System.out.print("Index " + i + ": ");
                Node curr = table[i];
                while (curr != null) 
                {
                    System.out.print("(" + curr.key + "," + curr.value + ") ");
                    curr = curr.next;
                }
                System.out.println();
            }
        }
    }
}

public class Main 
{
    public static void main(String[] args) 
    {
        Scanner sc = new Scanner(System.in);
        HashTable ht = new HashTable();
        while (true) 
        {
            int op = sc.nextInt();
            if (op == 1) 
            {
                int key = sc.nextInt();
                int value = sc.nextInt();
                ht.insert(key, value);
            } 
            else if (op == 2) 
            {
                int key = sc.nextInt();
                ht.delete(key);
            } 
            else if (op == 3) 
            {
                int key = sc.nextInt();
                ht.retrieve(key);
            } 
            else if (op == 4) 
            {
                ht.display();
            } 
            else if (op == 5) 
            {
                System.out.println("Exit");
                break;
            }
        }
    }
}

File: 20. Ch-5-3.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None

class HashTable:
    def __init__(self):
        self.size = 10
        self.table = [None] * self.size
    def hash_func(self, key):
        return key % self.size

    def insert(self, key, value):
        index = self.hash_func(key)
        new_node = Node(key, value)
        new_node.next = self.table[index]
        self.table[index] = new_node
        print(f"Inserted: ({key},{value})")

    def delete(self, key):
        index = self.hash_func(key)
        curr = self.table[index]
        prev = None
        while curr:
            if curr.key == key:
                if prev is None:
                    self.table[index] = curr.next
                else:
                    prev.next = curr.next
                curr.next=None
                print(f"Key {key} deleted successfully.")
                return
            prev = curr
            curr = curr.next
        print(f"Key {key} not found.")
    def retrieve(self, key):
        index = self.hash_func(key)
        curr = self.table[index]
        while curr:
            if curr.key == key:
                print(f"{key} -> {curr.value}")
                return
            curr = curr.next
        print(f"Key {key} not found.")

    def display(self):
        for i in range(self.size):
            if self.table[i]:
                print(f"Index {i}:", end=" ")
                curr = self.table[i]
                while curr:
                    print(f"({curr.key},{curr.value})", end=" ")
                    curr = curr.next
                print()

ht = HashTable()
while True:
    op = int(input())
    if op == 1:
        key, value = map(int, input().split())
        ht.insert(key, value)
    elif op == 2:
        key = int(input())
        ht.delete(key)
    elif op == 3:
        key = int(input())
        ht.retrieve(key)
    elif op == 4:
        ht.display()
    elif op == 5:
        print("Exit")
        break
